### PR TITLE
backend/local: do not retry epochNotMatch error when ingest sst

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve? 
- add retry for `SplitAndScatterRegionByRanges` since it may caused by region leader changes
- do not directly retry ingest if the return error is `EpochNotMatch` because tikv will check the epoch in SstMeta and the newest Region info, but we can change the epoch in SstMeta 
- retry ingest if the error is `Raft raft: proposal dropped`, because this means there is a leader transfer for this region
- downgrade some log level from Error to Warn, since they are retryable
- fix a bug in backend/log that cause scan region by range return one more region that does not intersect with the range  if the range end key is prefix of the region start key

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Side effects

Related changes

Release Note
- fix several trivial bugs when ingesting sst files in local backend

### CODE REVIEW

**Review by 3pointer — COMMENTED** [review]
LGTM

File: lightning/backend/local.go [hunk 1]
```diff
@@ -1211,65 +1234,32 @@ func (local *local) isIngestRetryable(
...
-	for i := len(key) - 1; i >= 0; i-- {
-		if key[i] != '\xff' {
-			pos = i
-			break
-		}
-	}
-	s, e := key[:pos], key[pos]+1
-	res = append(append(res, s...), e)
+	res = append(res, key...)
+	res = append(res, 0)
```
Comment (overvenus) [inline lightning/backend/local.go:1262]: Could you add a unit test?

File: lightning/backend/local.go [hunk 2]
```diff
@@ -1211,65 +1234,32 @@ func (local *local) isIngestRetryable(
...
-					}
-				}
-				if newLeader != nil {
-					newRegion = &split.RegionInfo{
-						Leader: newLeader,
-						Region: currentRegion,
-					}
-				}
-			}
+	case strings.Contains(errPb.Message, "Raft raft: proposal dropped"):
```
Comment (overvenus) [inline lightning/backend/local.go:1243]:
```suggestion
	case strings.Contains(errPb.Message, "raft: proposal dropped"):
```

Also please add some comments.

**Review by overvenus — COMMENTED** [review]
Rest LGTM

**Comment by overvenus** [comment]
Please add release note.

File: lightning/backend/local.go [hunk 3]
```diff
@@ -1202,7 +1209,23 @@ func (local *local) isIngestRetryable(
...
+			newRegion, err := local.splitCli.GetRegion(ctx, region.Region.GetStartKey())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if newRegion != nil {
+				return newRegion, nil
+			}
+			log.L().Warn("get region by key return nil, will retry", zap.Reflect("region", region),
+				zap.Int("retry", i))
+			time.Sleep(time.Second)
```
Comment (kennytm) [inline lightning/backend/local.go:1223]: select this `time.Sleep` with a `<-ctx.Done()` to avoid Ctrl+C being unresponsive

**Review by kennytm — COMMENTED** [review]
rest LGTM

File: lightning/backend/local.go [hunk 4]
```diff
@@ -1220,7 +1220,12 @@ func (local *local) isIngestRetryable(
 			}
 			log.L().Warn("get region by key return nil, will retry", zap.Reflect("region", region),
 				zap.Int("retry", i))
-			time.Sleep(time.Second)
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+				time.Sleep(time.Second)
```
Comment (kennytm) [inline lightning/backend/local.go:1227]:
```suggestion
			case <-time.After(time.Second):
```

**Review by overvenus — APPROVED** [review]
LGTM

**Pull request merged**